contents/ai/ml.md

```diff
@@ -0,0 +1,149 @@
+
+The distinction between the types of data is important as different ML algorithms perform better on different types of data. In particular, recent advances in neural networks has shown that they perform extremely well on unstructured data.
+
+## Machine Learning Workflow (Prototyping to Production)
+Bringing a machine learning algorithm in production requires a workflow which differs greatly from that of software engineering. This is due to a focus on prototyping. Prototyping is required as machine learning algorithms vary in efficacy when used in different domains. The quantity, quality and type of data differs greatly as well, sometimes containing errors such as corruption. Machine learning algorithms contain various parameters that can be tuned to improve performance.
+
+The prototyping phase in ML is usually done on [Jupyter notebooks](http://jupyter.org/), or [RStudio](https://www.rstudio.com/products/rstudio/). The highlight of such software is their strong support for visualization and quick code iterations. Visualizations are essential as they help practitioners analyze data and share findings with coworkers. Code style and clarity is less of a priority at this stage. 
+
+A machine learning algorithm is also refered to as a model after the algorithm is provided data to learn from.
+
+Prototyping can be broken down into the following phases:  
+1) Basic data preprocessing  
+2) Partitioning of data   
+3) Model training, evaluation and data analysis  
+4) Repeating step 3  
+
+### Basic Data Preprocessing
+Basic data preprocessing is required to remove unwanted noise from the data. Examples of unwanted noise are repeated content and corrupted images. The unwanted content is usually discarded. Training a model on such data hinders its ability to learn the actual patterns in uncorrupted data. 
+
+Statistics of the data is also collected to identify data imbalances. 
+
+![bar_chart](https://upload.wikimedia.org/wikipedia/commons/3/35/Incarceration_Rates_Worldwide_ZP.svg)  
+(image from wikipedia)
+
+Assume that the above barchart indicates the number of data points present for each country in the data. The number of data points for United States is much larger than that of Japan. Class imbalance is present.
+
+Class imbalance is known to negatively impact model performance. One easy remedey is oversampling the classes with fewer instances during training, up till the point where datapoints from each class has the same probability of being picked. This [paper](https://arxiv.org/abs/1710.05381v1) details oversampling and other methods for tackling class imbalance when using Convolutional Neural Networks. The literature review section in the paper also links to techniques effective on other machine learning algorithms.
+
+### Partitioning of Data
+Data is usually split into 3 sets, the training set, validation set and test set.
```

Comment:
I think you should talk about the sets in this order (training first!), rather than mentioning the test set first as seen below.

Reply:
You make the test set and validation set, and the training set is whatever that's left over so I ended up mentioning them in that order. Would swapping the order of the 3 in this line work better?